## Excercise 9.2

Create a subclass of `DiffMeansPermute` to perform a hypothesis test comparing means by resampling instead of permuting the current population.

In [39]:
import nsfg
import nsfg2
import first

import thinkstats2
import thinkplot

import numpy as np

class DiffMeansPermute(thinkstats2.HypothesisTest):
    """Tests a difference in means by permutation."""

    def TestStatistic(self, data):
        """Computes the test statistic.

        data: data in whatever form is relevant        
        """
        group1, group2 = data
        test_stat = abs(group1.mean() - group2.mean())
        #print(group1.mean())
        return test_stat

    def MakeModel(self):
        """Build a model of the null hypothesis.
        """
        group1, group2 = self.data
        self.n, self.m = len(group1), len(group2)
        self.pool = np.hstack((group1, group2))

    def RunModel(self):
        """Run the model of the null hypothesis.

        returns: simulated data
        """
        np.random.shuffle(self.pool)
        data = self.pool[:self.n], self.pool[self.n:]
        return data

class DiffMeansResample(DiffMeansPermute):
    def RunModel(self):
        sample1 = np.random.choice(self.pool, self.n, replace=True)
        sample2 = np.random.choice(self.pool, self.m, replace=True)
        return sample1, sample2

Calculate the new p-value using the resampling model comparing pregnancy lengths of the first birth versus subsequent births.

In [18]:
live, firsts, others = first.MakeFrames()
data = firsts.prglngth.values, others.prglngth.values
ht = DiffMeansResample(data)
p_val = ht.PValue()
print('p-value for pregnancy length',p_val)
print('actual difference of avg pregnancy length', ht.actual)

p-value for pregnancy length 0.161
actual difference 0.07803726677754952


The acutal difference in pregnancy length is 0.07 weeks with a $p$ value of 0.161. The $p$ value is large, and thus there is no real statistical significance between the average pregnancy length of a mother's first birth and any subsequent birth.

Now let's analyze the average birth weight of the baby comparing the first birth versus subsequent births.

In [41]:
data = firsts.totalwgt_lb.dropna().values, others.totalwgt_lb.dropna().values

ht = DiffMeansResample(data)
p_val = ht.PValue()
print('p-value for birth weight', p_val)
print('actual differnce of avg birth weight', ht.actual)

p-value for birth weight 0.0
actual differnce of average birth weight 0.12476118453549034


The actual difference in average birth weight is 0.12 lbs with a $p$ value of 0.0. There is a statistical difference between the average birth weight of the first birth and all subsequent births.